### PR TITLE
feat: add java/dotnet workshop

### LINKED ISSUE #72: Java Workshop
### General Information

* [x] :wave: I may be able to implement this feature request

* [ ] :warning: This feature might incur a breaking change 

### Description

We should implement a Java version of the workshop procedure to match GA languages for CDK

### PR DESCRIPTION
**Pull request by NGL321**
- Adds C# workshop
- Adds Java workshop
- Fixes exposed ARN in python workshop

Fixes #72 

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: workshop/content/40-dotnet/20-create-project/300-structure.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+using System.Collections.Generic;
+using System.Linq;
+
+namespace CdkWorkshop
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            var app = new App(null);
```
Comment (RomainMuller) [inline workshop/content/40-dotnet/20-create-project/300-structure.md]: Remove the unneeded `null` arguments everywhere :P (You may want to re-generate the samples from the fixed up init templates once those have merged :D)

File: workshop/content/40-dotnet/20-create-project/300-structure.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            new CdkWorkshopStack(app, "CdkWorkshopStack", new StackProps());
```
Comment (RomainMuller) [inline workshop/content/40-dotnet/20-create-project/300-structure.md]: Also here - the last argument shouldn't be there anymore.

File: workshop/content/40-dotnet/30-hello-cdk/200-lambda.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+namespace CdkWorkshop
+{
+    public class CdkWorkshopStack : Stack
+    {
+        public CdkWorkshopStack(Construct scope, string id, IStackProps props) : base(scope, id, props)
+        {
+            // Defines a new lambda resource
+            var hello = new Function(this, "HelloHandler", new FunctionProps()
+            {
+                Runtime = Runtime.NODEJS_10_X, // execution environment
```
Comment (RomainMuller) [inline workshop/content/40-dotnet/30-hello-cdk/200-lambda.md]: <3

File: workshop/content/40-dotnet/30-hello-cdk/300-apigw.md [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+{{<highlight ts "hl_lines=3 19-23">}}
+using Amazon.CDK;
+using Amazon.CDK.AWS.Lambda;
+using Amazon.CDK.AWS.APIGateway;
+
+namespace CdkWorkshop
+{
+    public class CdkWorkshopStack : Stack
+    {
+        public CdkWorkshopStack(Construct parent, string id, IStackProps props) : base(parent, id, props)
```
Comment (RomainMuller) [inline workshop/content/40-dotnet/30-hello-cdk/300-apigw.md]: Make sure the `props` argument has a `= null` default (same here - updated template will have that).

File: workshop/content/40-dotnet/30-hello-cdk/200-lambda.md [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+## Add an AWS Lambda Function to your stack
+
+Add an `using` statement at the beginning of `src/CdkWorkshop/CdkWorkshopStack.cs`, and a
+`Function` to your stack.
+
+
+{{<highlight csharp "hl_lines=2 10-16">}}
+using Amazon.CDK;
+using Amazon.CDK.AWS.Lambda;
+
```
[the remark this answers is not in the document]
Comment (nija-at) [reply workshop/content/40-dotnet/30-hello-cdk/200-lambda.md]: Romain said that some were due to dotnet code being marked as typescript, which he has fixed. Let's do another pass to check on this though.

Some other places, we have code blocks vs screenshots, and they're inconsistent as well.

File: workshop/content/40-dotnet/20-create-project/300-structure.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        }
+    }
+}
+
+```
+
+This code loads and instantiate the `CdkWorkshopStack` class from the
+`src/CdkWorkshop/CdkWorkshopStack.cs` file. We won't need to look at this file anymore.
+
+## The main stack
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply workshop/content/40-dotnet/20-create-project/300-structure.md]: Template has changed (new version yet to be released)

File: workshop/content/40-dotnet/20-create-project/400-synth.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+application.
+
+To synthesize a CDK app, use the `cdk synth` command. Let's check out the
+template synthesized from the sample app:
+
+```
+cdk synth
+```
+
+Will output the following CloudFormation template:
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply workshop/content/40-dotnet/20-create-project/400-synth.md]: I don't think this is true... The *old* template definitely had two stacks, the upcoming one won't.

See aws/aws-cdk#5119
See aws/aws-cdk#5144

**Comment by RomainMuller** [comment]
Note - we need to coordinate deploying the updated website **AFTER** the CDK release that includes aws/aws-cdk#5119 and aws/aws-cdk#5144, otherwise templates will not match!

File: workshop/content/40-dotnet/20-create-project/500-deploy.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
++++
+title = "cdk deploy"
+weight = 500
++++
+
+Okay, we've got a CloudFormation template. What's next? __Let's deploy it into our account!__
+
+## Bootstrapping an environment
```
[the remark this answers is not in the document]
Comment (NGL321) [reply workshop/content/40-dotnet/20-create-project/500-deploy.md]: That is done in the "Prerequisites" section of the workshop

File: workshop/content/40-dotnet/30-hello-cdk/100-cleanup.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+what's currently deployed. This is a safe way to check what will happen once we run `cdk deploy` and is always good practice:
+
+```
+cdk diff
+```
+
+Output should look like the following:
+
+```
+IAM Statement Changes
```
[the remark this answers is not in the document]
Comment (NGL321) [reply workshop/content/40-dotnet/30-hello-cdk/100-cleanup.md]: I have tried a number of things to fix this. I believe it is always going to look a little awkward on narrow screens, but on anything normal width it should be fine.

File: workshop/content/40-dotnet/20-create-project/300-structure.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+![](./structure.png)
+
+//TODO
+
+* `src/CdkWorkshop/Program.cs` is the entrypoint for the CDK application it will load the stack defined in `src/CdkWorkshop/CdkWorkshopStack.cs`
+* `src/CdkWorkshop/CdkWorkshopStack.cs` is where your CDK application's main stack is defined. This is the file we'll be spending most of our time in.
+* `cdk.json` tells the toolkit how to run your app. In our case it will be `"dotnet run -p src/CdkWorkshop/CdkWorkshop.csproj"`
+* `src/CdkWorkshop/CdkWorkshop.csproj` is the C# project file. It is an xml file and contains information on references. This will be useful to you down the line, but is not relevant for the purposes of this workshop.
+* `src/CdkWorkshop/GlobalSuppressions.cs` disables the Roslyn analyzer for `RECS0026:Possible unassigned object created by 'new'` as this generates many false positives with CDK.
```
[the remark this answers is not in the document]
Comment (NGL321) [reply workshop/content/40-dotnet/20-create-project/300-structure.md]: Should be fixed. Shows working on demo site

File: workshop/content/40-dotnet/20-create-project/300-structure.md [hunk 11]
```diff
@@ -0,0 +1,98 @@
++++
+title = "Project structure"
+weight = 300
++++
+
+## Open your IDE
+
+Now's a good time to open the project in your favorite IDE and explore.
+
+> If you use VSCode, you can just type `code .` within the project directory.
```
[the remark this answers is not in the document]
Comment (NGL321) [reply workshop/content/40-dotnet/20-create-project/300-structure.md]: Fixed

File: workshop/content/40-dotnet/20-create-project/300-structure.md [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+![](./structure.png)
+
+//TODO
+* `src/CdkWorkshop/Program.cs` is the entrypoint for the CDK application it will load the stack defined in `src/CdkWorkshop/CdkWorkshopStack.cs`
+* `src/CdkWorkshop/CdkWorkshopStack.cs` is where your CDK application's main stack is defined. This is the file we'll be spending most of our time in.
+* `cdk.json` tells the toolkit how to run your app. In our case it will be
+  `"node bin/cdk-workshop.js"`
+* `src/CdkWorkshop/CdkWorkshop.csproj` is the C# project file. It is an xml file and contains information on references. This will be useful to you down the line, but is not relevant for the purposes of this workshop.
+* `src/CdkWorkshop.sln` is the C# solution file that provides build information. You should not need to interface with this file.
+* `.gitignore` and `.npmignore` tell git and npm which files to include/exclude
```
[the remark this answers is not in the document]
Comment (NGL321) [reply workshop/content/40-dotnet/20-create-project/300-structure.md]: Fixed

File: workshop/content/40-dotnet/20-create-project/300-structure.md [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+You'll see something like this:
+
+![](./structure.png)
+
+//TODO
+* `src/CdkWorkshop/Program.cs` is the entrypoint for the CDK application it will load the stack defined in `src/CdkWorkshop/CdkWorkshopStack.cs`
+* `src/CdkWorkshop/CdkWorkshopStack.cs` is where your CDK application's main stack is defined. This is the file we'll be spending most of our time in.
+* `cdk.json` tells the toolkit how to run your app. In our case it will be
+  `"node bin/cdk-workshop.js"`
```
[the remark this answers is not in the document]
Comment (NGL321) [reply workshop/content/40-dotnet/20-create-project/300-structure.md]: Fixed

File: workshop/content/50-java/20-create-project/300-structure.md [hunk 14]
```diff
@@ -22,7 +22,7 @@ You'll see something like this:
   `"mvn -q exec:java"`
 * `pom.xml` is the Maven project file. It is an xml file and contains information on build properties, dependencies, and app information. This will be useful to you down the line, but is not relevant for the purposes of this workshop.
 * `test/java/com/myorg/CdkWorkshopStackTest.java` Is a build test that is run againt the Java application on build and will indicated success or failure in the terminal. We will not be using this for the purposes of the workshop.
-* `.gitignore` and `.npmignore` tell git and npm which files to include/exclude
+* `.gitignore` tell git and npm which files to include/exclude
```
Comment (shivlaks) [inline workshop/content/50-java/20-create-project/300-structure.md]: drop the `and npm part`

File: workshop/content/40-dotnet/50-table-viewer/400-expose-table.md [hunk 15]
```diff
@@ -74,7 +74,7 @@ namespace CdkWorkshop
 {
     public class CdkWorkshopStack : Stack
     {
-        public CdkWorkshopStack(Construct scope, string id, IStackProps props = null) : base(scope, id, props)
+        public CdkWorkshopStack(Construct scope, string id) : base(scope, id)
```
Comment (shivlaks) [inline workshop/content/40-dotnet/50-table-viewer/400-expose-table.md]: was this intentional? I thought @RomainMuller wanted this... it's in our init templates

File: workshop/content/40-dotnet/20-create-project/300-structure.md [hunk 16]
```diff
@@ -8,22 +8,24 @@ weight = 300
...
-//TODO
-
 * `src/CdkWorkshop/Program.cs` is the entrypoint for the CDK application it will load the stack defined in `src/CdkWorkshop/CdkWorkshopStack.cs`
 * `src/CdkWorkshop/CdkWorkshopStack.cs` is where your CDK application's main stack is defined. This is the file we'll be spending most of our time in.
 * `cdk.json` tells the toolkit how to run your app. In our case it will be `"dotnet run -p src/CdkWorkshop/CdkWorkshop.csproj"`
 * `src/CdkWorkshop/CdkWorkshop.csproj` is the C# project file. It is an xml file and contains information on references. This will be useful to you down the line, but is not relevant for the purposes of this workshop.
 * `src/CdkWorkshop/GlobalSuppressions.cs` disables the Roslyn analyzer for `RECS0026:Possible unassigned object created by 'new'` as this generates many false positives with CDK.
 * `src/CdkWorkshop.sln` is the C# solution file that provides build information. You should not need to interface with this file.
-* `.gitignore` and `.npmignore` tell git and npm which files to include/exclude
+* `.gitignore` tell git and npm which files to include/exclude
```
Comment (shivlaks) [inline workshop/content/40-dotnet/20-create-project/300-structure.md]: drop the `and npm` part

File: workshop/content/40-dotnet/20-create-project/300-structure.md [hunk 17]
```diff
@@ -67,7 +69,7 @@ namespace CdkWorkshop
 {
     public class CdkWorkshopStack : Stack
     {
-        public CdkWorkshopStack(Construct parent, string id, IStackProps props = null) : base(parent, id, props)
+        public CdkWorkshopStack(Construct scope, string id) : base(scope, id)
```
Comment (shivlaks) [inline workshop/content/40-dotnet/20-create-project/300-structure.md]: revert?

File: workshop/content/40-dotnet/30-hello-cdk/100-cleanup.md [hunk 18]
```diff
@@ -18,7 +18,7 @@ namespace CdkWorkshop
 {
     public class CdkWorkshopStack : Stack
     {
-        public CdkWorkshopStack(Construct scope, string id, IStackProps props = null) : base(scope, id, props)
+        public CdkWorkshopStack(Construct scope, string id) : base(scope, id)
```
Comment (shivlaks) [inline workshop/content/40-dotnet/30-hello-cdk/100-cleanup.md]: ditto

File: workshop/content/40-dotnet/30-hello-cdk/300-apigw.md [hunk 19]
```diff
@@ -32,7 +32,7 @@ namespace CdkWorkshop
 {
     public class CdkWorkshopStack : Stack
     {
-        public CdkWorkshopStack(Construct parent, string id, IStackProps props = null) : base(parent, id, props)
+        public CdkWorkshopStack(Construct scope, string id) : base(scope, id)
```
Comment (shivlaks) [inline workshop/content/40-dotnet/30-hello-cdk/300-apigw.md]: ditto

File: workshop/content/40-dotnet/50-table-viewer/300-add.md [hunk 20]
```diff
@@ -19,7 +19,7 @@ namespace CdkWorkshop
 {
     public class CdkWorkshopStack : Stack
     {
-        public CdkWorkshopStack(Construct scope, string id, IStackProps props = null) : base(scope, id, props)
+        public CdkWorkshopStack(Construct scope, string id) : base(scope, id)
```
Comment (shivlaks) [inline workshop/content/40-dotnet/50-table-viewer/300-add.md]: ditto

File: workshop/content/40-dotnet/20-create-project/300-structure.md (hunk 17 above)
Comment (NGL321) [reply workshop/content/40-dotnet/20-create-project/300-structure.md]: Fixed

File: workshop/content/50-java/20-create-project/300-structure.md (hunk 14 above)
Comment (NGL321) [reply workshop/content/50-java/20-create-project/300-structure.md]: Fixed

File: workshop/content/40-dotnet/50-table-viewer/400-expose-table.md (hunk 15 above)
Comment (NGL321) [reply workshop/content/40-dotnet/50-table-viewer/400-expose-table.md]: Change made before the discussion. Fixed

File: workshop/content/40-dotnet/30-hello-cdk/100-cleanup.md (hunk 18 above)
Comment (NGL321) [reply workshop/content/40-dotnet/30-hello-cdk/100-cleanup.md]: fixed

File: workshop/content/40-dotnet/20-create-project/300-structure.md (hunk 16 above)
Comment (NGL321) [reply workshop/content/40-dotnet/20-create-project/300-structure.md]: fixed

File: workshop/content/40-dotnet/50-table-viewer/300-add.md (hunk 20 above)
Comment (NGL321) [reply workshop/content/40-dotnet/50-table-viewer/300-add.md]: fixed

File: workshop/content/40-dotnet/30-hello-cdk/300-apigw.md (hunk 19 above)
Comment (NGL321) [reply workshop/content/40-dotnet/30-hello-cdk/300-apigw.md]: fixed

**Pull request merged**